controller/src/lib.rs

```diff
@@ -468,15 +562,27 @@ impl IoLoop {
                     );
                     self.send_outgoing_request(&mut tx_buf).await;
                     self.resend.reset();
+                    n_retries += 1;
                 }
 
                 // If we _do_ have an outstanding request, we need to resend it
                 // periodically until we get a response. Wait for up to the resend
                 // interval of inactivity, and then possibly retry.
                 _ = self.resend.tick(), if self.outstanding_request.is_some() => {
-                    debug!(self.log, "timed out without response, retrying");
-                    self.send_outgoing_request(&mut tx_buf).await;
-                    self.resend.reset();
+                    if n_retries < self.n_retries {
+                        debug!(self.log, "timed out without response, retrying");
+                        self.send_outgoing_request(&mut tx_buf).await;
```

Comment:
Nit - Would it make sense for `send_outgoing_request` to call `self.resend.reset()` and/or increment `n_retries` instead of duplicating this between here and line 563-565 above? `n_retries` is currently a local var so that might be awkward (unless it could reasonably be a field on `self`? not sure about that).

Reply:
I think I tried that but found it awkward since `send_outgoing_request` takes a shared rather than exclusive reference. But there's no reason it can't be modified to take `&mut self`! I'll add the number of retries to the type as well, and do those modifications in that method. Thanks!